internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf/probe.bpf.c

```diff
@@ -0,0 +1,196 @@
+        bpf_printk("write_span_context: failed to write trace flags: %ld", ret);
+        return -4;
+    }
+
+    return 0;
+}
+
+// This instrumentation attaches a uprobe to the following function:
+// func (t *tracer) start(ctx context.Context, spanPtr *span, parentSpanCtx *trace.SpanContext, sampled *bool, spanCtx *trace.SpanContext) {
+// https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/effdec9ac23e56e9e9655663d386600e62b10871/sdk/trace.go#L56-L66
+SEC("uprobe/Tracer_start")
+int uprobe_Tracer_start(struct pt_regs *ctx) {
+    struct go_iface go_context = {0};
+    get_Go_context(ctx, 2, 0, true, &go_context);
+
+    struct otel_span_t otel_span;
+    __builtin_memset(&otel_span, 0, sizeof(struct otel_span_t));
+
+    start_span_params_t params = {
+        .ctx = ctx,
+        .go_context = &go_context,
+        .psc = &otel_span.psc,
+        .sc = &otel_span.sc,
+        .get_parent_span_context_fn = NULL,
+        .get_parent_span_context_arg = NULL, // Default to new root. 
+    };
+
+    start_span(&params);
+
+    void *parent_span_context = get_argument(ctx, 5);
+    long rc = write_span_context(parent_span_context, &otel_span.psc);
+    if (rc != 0) {
+        bpf_printk("failed to write parent span context: %ld", rc);
+    }
+
+    if (!is_sampled(params.sc)) {
+        // Default SDK behaviour is to sample everything. Only set the sampled
+        // value to false when needed.
+        void *sampled_ptr_val = get_argument(ctx, 6);
+        if (sampled_ptr_val == NULL) {
+            bpf_printk("nil sampled pointer");
+        } else {
+            bool false_val = false;
+            rc = bpf_probe_write_user(sampled_ptr_val, &false_val, sizeof(bool));
+            if (rc != 0) {
+                bpf_printk("bpf_probe_write_user: failed to write sampled value: %ld", rc);
+            } else {
+                bpf_printk("wrote sampled value");
+            }
+        }
+    }
+
+    void *span_context_ptr_val = get_argument(ctx, 7);
+    rc = write_span_context(span_context_ptr_val, &otel_span.sc);
+    if (rc != 0) {
+        bpf_printk("failed to write span context: %ld", rc);
+    }
+
+    void *span_ptr_val = get_argument(ctx, 4);
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, &otel_span, 0);
+    start_tracking_span(go_context.data, &otel_span.sc);
+
+    return 0;
+}
+
+// This instrumentation attaches a uprobe to the following function:
+// func (*span) ended(buf []byte) {}
+// https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/effdec9ac23e56e9e9655663d386600e62b10871/sdk/trace.go#L133-L136
+SEC("uprobe/Span_ended")
+int uprobe_Span_ended(struct pt_regs *ctx) {
+    void *span_ptr = get_argument(ctx, 1);
+    struct otel_span_t *span = bpf_map_lookup_elem(&active_spans_by_span_ptr, &span_ptr);
+    if (span == NULL) {
+        return 0;
+    }
+    stop_tracking_span(&span->sc, &span->psc);
+    bpf_map_delete_elem(&active_spans_by_span_ptr, &span_ptr);
```

Comment:
We should call delete only once we finish any handling with the underlying memory of the map value.

Reply:
I think the only thing we do with the read `span` is check sampling. I'll update to store that in a local var prior to deleting.

Reply:
Updated. PTAL.